README.md
# OWASP jOTP

[![Build Status](https://travis-ci.org/upcrob/jOTP.png)](https://travis-ci.org/upcrob/jOTP)
[![Download](https://api.bintray.com/packages/upcrob/generic/jOTP/images/download.svg)](https://bintray.com/upcrob/generic/jOTP/_latestVersion)

[OWASP jOTP](https://www.owasp.org/index.php/OWASP_JOTP_Project) is a microservice for generating and validating one-time passwords as a secondary factor of authentication.

A common use case for jOTP is as follows:

1. Client applications displays a login page requesting the user enter his/her username and password.
2. If the credentials check passes, the user's email is looked up and a message containing the token is sent.
3. The application then requests that the OTP token that was sent be entered in a text box.  Once entered, it is sent to jOTP.
4. jOTP validates the token.  If the token was valid, the application finishes authenticating the user.  If the token was not valid, the user is redirected to the login page.

## Usage

### Monitor Function

**Description:** The monitor function simply lets the caller know whether or not the application is available.  It will return an HTTP 200 status and "OK" in the response body if jOTP is available.

**HTTP Verb:** GET

**URL:** `/`

**Parameters:** None

### Generate OTP Token

**Description:** POSTing to the `/sessions` endpoint tells jOTP to generate an OTP token and send it to the user.  If successful, jOTP responds with a 201 and the session endpoint to validate against in the `Location` header.

**HTTP Verb:** POST

**URL:** `/sessions`

**Request Body Format:**

	{
		"email": "[email address to send token to]",
		"subject": "[email subject]",
		"message": "[email message]",
		"ttl": "[number of minutes before token expires]"
	}

jOTP replaces `$token` in the message with the generated token and `$session` with the generated session.

### Validate Token

**Description:** POSTing the OTP token to the matching session endpoint validates the token.  jOTP responds with a 200 HTTP status if the token is valid.

**HTTP Verb:** POST

**URL:** `/sessions/[session id]`

**Request Body:** [plaintext OTP token]

## Configuration

The following 2 JVM properties are used by the application.  These can be customized by setting the `JAVA_OPTS` environment variable.:

* **propertyfiles** This should be a (comma-delimited) list of property files that jOTP should read when it starts up.  This defaults to a file called `jotp.properties` located in the current working directory.
* **logfile** This should be the absolute path of the file where jOTP logs should be written.  Log files will be rolled over hourly in the same directory.  This defaults to a file called `jotp.log` located in the current working directory.

The following properties may be set in any of the property files specified in the `properyfiles` JVM property:

**log.file** Path to log file.  Files will be rolled over with the date in the same directory.

**http.port** HTTP port the web server should listen on.

**base.url** The base URL for the service (e.g. http://localhost:8080).  This is used to construct the `Location` header.

**smtp.host** SMTP server.

**smtp.port** SMTP server port.

**smtp.from** Email address to send emails from.

**smtp.tls** If jOTP should connect to the SMTP server over TLS.

**smtp.username** SMTP username.  Optional.

**smtp.password** SMTP password.  Optional.

**repository.type** Type of repository to use (`redis` or `jdbc`).  If not set, jOTP defaults to a local in-memory repository that is useful for local development but not recommended for production environments.

### Redis Properties

**redis.host** Redis server address.

**redis.port** Redis server port.

**redis.creds** Redis server credentials.  Optional.

### JDBC Properties

**jdbc.url** URL for connecting to DBMS.  Note that jOTP does not include any JDBC drivers so these will need to be added to your runtime classpath.

**jdbc.table** Name of the table jOTP should use to store tokens in.  It should follow the following schema:

	session    VARCHAR(255) PRIMARY KEY
	token      VARCHAR(255) PRIMARY KEY
	expiretime TIMESTAMP


RELEASENOTES.md
# jOTP Release Notes

## v2.0.0

* Simplified jOTP architecture and configuration.

## v1.0.0

* Updated wrapper for Gradle 2.3.
* Added Gradle Bintray integration.

## v0.4.0

* Updated Redis Tokenstore to use built-in expiration.
* Removed passwords from debug log.

## v0.3.2

* Changed project to use OWASP namespace.

## v0.3.1

* Updated Gradle mail dependencies.
* Added .gitignore

## v0.3.0

* Added configuration options for SMTP TLS and authentication.

## v0.2.0

* First 'officially' released version.


src/main/java/org/owasp/jotp/Main.java
package org.owasp.jotp;

import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;
import java.util.Properties;
import java.util.UUID;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import spark.Request;
import spark.Response;
import spark.Spark;

import com.google.gson.*;

import static spark.Spark.*;

/**
 * jOTP driver class.
 */
public class Main {
	
	private static final Gson gson = new Gson();
	private static final Logger logger = LoggerFactory.getLogger(Main.class);
	
	private static SessionRepository repository;
	private static EmailSender emailSender;
	private static String baseUrl;
	
	private static Map<String, String> properties;
	
	public static void main(String[] args) {
		// load property files
		String propertyFiles = System.getProperty("propertyfiles");
		if (propertyFiles == null || "".equals(propertyFiles.trim())) {
			System.err.println("No property files specified.  Set the 'propertyfiles' property for the VM.");
			System.exit(1);
		}
		properties = new HashMap<>();
		String[] files = propertyFiles.split(",");
		for (String file : files) {
			Properties props = new Properties();
			try {
				props.load(new FileReader(new File(file)));
				for (Object k : props.keySet()) {
					properties.put(k.toString(), props.get(k).toString());
				}
			} catch (IOException e) {
				System.err.println("Failed to load property file: "
						+ file + "  " + e.getMessage());
				System.exit(1);
			}
		}
		
		// configure
		int port = getPropertyAsInt("http.port", true);
		baseUrl = getPropertyAsString("base.url", true);
		String smtpHost = getPropertyAsString("smtp.host", true);
		int smtpPort = getPropertyAsInt("smtp.port", true);
		String smtpFrom = getPropertyAsString("smtp.from", true);
		boolean smtpTls = getPropertyAsBoolean("smtp.tls", true);
		String keystoreFile = getPropertyAsString("keystore.file", false);
		String keystorePassword = getPropertyAsString("keystore.password", false);
		String truststoreFile = getPropertyAsString("truststore.file", false);
		String truststorePassword = getPropertyAsString("truststore.password", false);
		
		// configure repository
		String repoType = getPropertyAsString("repository.type", false);
		if ("redis".equals(repoType)) {
			logger.info("Configuring Redis repository.");
			repository = new RedisSessionRepository(getPropertyAsString("redis.host", true),
					getPropertyAsInt("redis.port", true),
					getPropertyAsString("redis.creds", false));
		} else if ("jdbc".equals(repoType)) {
			logger.info("Configuring JDBC repository.");
			repository = new JdbcSessionRepository(getPropertyAsString("jdbc.url", true),
					getPropertyAsString("jdbc.table", true));
		} else {
			logger.warn("Using local in-memory repository.  This is not recommended for production environments.");
			repository = new InMemorySessionRepository();
		}
		
		// configure sender
		String smtpUsername = getPropertyAsString("smtp.username", false);
		String smtpPassword = getPropertyAsString("smtp.password", false);
		if (smtpUsername != null && smtpPassword != null) {
			emailSender = new EmailSender(smtpHost, smtpPort, smtpFrom, smtpTls, smtpUsername, smtpPassword);
		} else {
			emailSender = new EmailSender(smtpHost, smtpPort, smtpFrom, smtpTls);
		}
		
		// configure shutdown hook to close repository
		Runtime.getRuntime().addShutdownHook(new Thread() {
			@Override
			public void run() {
				logger.info("Shutting down web server...");
				Spark.stop();
				logger.info("Server stopped.");
				logger.info("Shutting down repository...");
				repository.close();
				logger.info("Shutdown complete.");
			}
		});
		
		// configure http server
		port(port);
		if (keystoreFile != null && keystorePassword != null
				&& truststoreFile != null && truststorePassword != null) {
			secure(keystoreFile, keystorePassword, truststoreFile, truststorePassword);
		}
		post("/sessions", (req, res) -> handleTokenGenerationRequest(req, res));
		post("/sessions/:session", (req, res) -> handleTokenValidationRequest(req, res));
		get("/", (req, res) -> handleMonitorRequest(req, res));
	}

	private static String handleTokenGenerationRequest(Request req, Response res) {
		SessionGenerationRequest sessReq = gson.fromJson(req.body(), SessionGenerationRequest.class);
		
		String token = randomString().substring(0, 5).toUpperCase();
		String session = randomString().substring(0, 16);
		try {
			repository.createSession(session, token, LocalDateTime.now().plusMinutes(sessReq.ttl));
			emailSender.send(sessReq.email, sessReq.subject, replaceMessageVars(sessReq.message, session, token));
			res.status(201);
			res.header("Location", baseUrl + "/sessions/" + session);
			return "";
		} catch (ServiceException e) {
			logger.error("Session repository error during session creation.", e);
			res.status(500);
			return "Internal Server Error";
		}
	}
	
	private static String handleTokenValidationRequest(Request req, Response res) {
		String session = req.params(":session");
		String token = req.body();
		try {
			if (repository.isValid(session, token)) {
				res.status(200);
			} else {
				res.status(404);
			}
		} catch (ServiceException e) {
			res.status(500);
			logger.error("Session repository error during validation.", e);
		}
		
		return "";
	}

	private static String handleMonitorRequest(Request req, Response res) {
		res.status(200);
		return "OK";
	}
	
	private static String getPropertyAsString(String name, boolean required) {
		String value = properties.get(name);
		if (value == null) {
			if (required)
				throw new IllegalArgumentException("Missing configuration property: " + name);
		}
		logger.info("Got configuration property: " + name);
		return value;
	}
	
	private static int getPropertyAsInt(String name, boolean required) {
		String value = properties.get(name);
		if (value == null) {
			if (required)
				throw new IllegalArgumentException("Missing configuration property: " + name);
		} else if (!value.matches("[0-9]+")) {
			throw new IllegalArgumentException("Invalid value for configuration property: " + name);
		}
		logger.info("Got configuration property: " + name);
		return Integer.parseInt(value);
	}
	
	private static boolean getPropertyAsBoolean(String name, boolean required) {
		String value = properties.get(name);
		if (value == null) {
			if (required)
				throw new IllegalArgumentException("Missing configuration property: " + name);
		}
		boolean boolValue = Boolean.valueOf(value).booleanValue();
		logger.info("Got configuration property: " + name);
		return boolValue;
	}
	
	private static String randomString() {
		UUID uuid = UUID.randomUUID();
		return uuid.toString().replaceAll("-", "");
	}
	
	private static String replaceMessageVars(String message, String session, String token) {
		return message.replace("$session", session).replace("$token", token);
	}
}
